lib/util/package-rules/package-names.ts

```diff
@@ -4,28 +4,28 @@ import { Matcher } from './base';
 
 export class PackageNameMatcher extends Matcher {
   override matches(
-    { depName }: PackageRuleInputConfig,
+    { depName, packageName }: PackageRuleInputConfig,
     { matchPackageNames }: PackageRule
   ): boolean | null {
     if (is.undefined(matchPackageNames)) {
       return null;
     }
-    if (is.undefined(depName)) {
-      return false;
-    }
-    return matchPackageNames.includes(depName);
+    return (
+      (is.string(depName) && matchPackageNames.includes(depName)) ||
+      (is.string(packageName) && matchPackageNames.includes(packageName))
+    );
   }
 
   override excludes(
-    { depName }: PackageRuleInputConfig,
+    { depName, packageName }: PackageRuleInputConfig,
     { excludePackageNames }: PackageRule
   ): boolean | null {
     if (is.undefined(excludePackageNames)) {
       return null;
     }
-    if (is.undefined(depName)) {
-      return false;
-    }
-    return excludePackageNames.includes(depName);
+    return (
+      (is.string(depName) && excludePackageNames.includes(depName)) ||
```

Comment:
Need to think if this should be and, not or. Otherwise this will always exclude if depName != packageName

Reply:
What do you mean by _always_ exclude?
As I see it, It will only exclude if one of these is set by a user's package rule and there is a match.

<details><summary>All these tests are currently passing (with or) </summary>

```ts
describe('excludePackageNames', () => {
    it('does not apply excludePackageNames (depName==packageName)', () => {
      const config: TestConfig = {
        depName: 'xyz/bar',
        packageName: 'xyz/bar',
        packageRules: [
          {
            matchPackagePrefixes: ['xyz/'],
            excludePackageNames: ['xyz/foo'],
            groupName: 'xyz',
          },
        ],
      };
      const res = applyPackageRules(config);
      expect(res.groupName).toBe('xyz');
    });

    it('applies excludePackageNames (depName==packageName)', () => {
      const config: TestConfig = {
        depName: 'xyz/foo',
        packageName: 'xyz/foo',
        packageRules: [
          {
            matchPackagePrefixes: ['xyz/'],
            excludePackageNames: ['xyz/foo'],
            groupName: 'xyz',
          },
        ],
      };
      const res = applyPackageRules(config);
      expect(res.groupName).toBeUndefined();
    });

    it('does not apply excludePackageNames (depName!=packageName)', () => {
      const config: TestConfig = {
        depName: 'xyz/bar',
        packageName: 'uvw:xyz/bar',
        packageRules: [
          {
            matchPackagePrefixes: ['xyz/'],
            excludePackageNames: ['xyz/foo'],
            groupName: 'xyz',
          },
        ],
      };
      const res = applyPackageRules(config);
      expect(res.groupName).toBe('xyz');
    });

    it('also applies excludePackageNames (depName!=packageName)', () => {
      const config: TestConfig = {
        depName: 'xyz/foo',
        packageName: 'uvw:xyz/foo',
        packageRules: [
          {
            matchPackagePrefixes: ['xyz/'],
            excludePackageNames: ['xyz/foo'],
            groupName: 'xyz',
          },
        ],
      };
      const res = applyPackageRules(config);
      expect(res.groupName).toBeUndefined();
    });
  });

```

</details>

Setting it to "and" will cause some test to fail.
And if we indeed want to got for "and", than we should change the `excludePackagePattern` to "and" as well.
But I think going for "and" will break existing user configurations, so it might be a breaking change (?)

Reply:
And i afraid we missed `matchPackagePrefixes` and `excludePackagePrefixes` 😅, which also only match on `depName`.

ill handle these as well.

Reply:
https://github.com/renovatebot/renovate/pull/19722